R/diagnose_design.R

```diff
@@ -177,104 +217,301 @@ diagnose_design_single_design <- function(design, diagnosands, sims, bootstrap)
+#' @param diagnosands A set of diagnosands created by \code{\link{declare_diagnosands}}. By default, these include bias, root mean-squared error, power, frequentist coverage, the mean and standard deviation of the estimate(s), the "type S" error rate (Gelman and Carlin 2014), and the mean of the estimand(s).
+#' @param add_parameters If TRUE, add any design parameters to the diagnosands_df and the bootrap_df, inherited from simulation_df (and ultimately, from designs)
+#' @param add_grouping_variables Variables used to generate groups of simulations for diagnosis. Added to list default list: c("design_ID", "estimand_label", "estimator_label", "coefficient")
+#'
+#' @param sims The number of simulations, defaulting to 500. sims may also be a vector indicating the number of simulations for each step in a design, as described for \code{\link{simulate_design}}
+#' @param bootstrap Number of bootstrap replicates for the diagnosands to obtain the standard errors of the diagnosands, defaulting to \code{100}.
+#'
+#' @details
+#'
+#' If the diagnosand function contains a \code{group_by} attribute, it will be used to split-apply-combine diagnosands rather than the intersecting column names.
+#'
+#' If \code{sims} is named, or longer than one element, a fan-out strategy is created and used instead.
+#'
+#' If the \code{future} package is installed, you can set  \code{\link[future]{plan}} to run multiple simulations at once.
+#'
+#'
+#' @examples
+#' my_population <- declare_population(N = 500, noise = rnorm(N))
+#'
+#' my_potential_outcomes <- declare_potential_outcomes(
+#'   Y_Z_0 = noise, Y_Z_1 = noise +
+#'   rnorm(N, mean = 2, sd = 2))
+#'
+#' my_assignment <- declare_assignment()
+#'
+#' my_estimand <- declare_estimand(ATE = mean(Y_Z_1 - Y_Z_0))
+#'
+#' my_estimator <- declare_estimator(Y ~ Z, estimand = my_estimand)
+#'
+#' my_reveal <- declare_reveal()
+#'
+#' design <- declare_design(my_population,
+#'                          my_potential_outcomes,
+#'                          my_estimand,
+#'                          my_assignment,
+#'                          my_reveal,
+#'                          my_estimator)
+#'
+#' \dontrun{
+#' # using built-in defaults:
+#' diagnosis <- diagnose_design(design)
+#' diagnosis
+#' }
+#'
+#' # using a user-defined diagnosand
+#' my_diagnosand <- declare_diagnosands(absolute_error = mean(abs(est - estimand)))
+#'
+#' \dontrun{
+#' diagnosis <- diagnose_design(design, diagnosands = my_diagnosand)
+#' diagnosis
+#'
+#' get_diagnosands(diagnosis)
+#'
+#' get_simulations(diagnosis)
+#'
+#' }
+#' # Using an existing data frame of simulations
+#' \dontrun{
+#' simulations <- simulate_design(designs, sims = 2)
+#' diagnosis   <- diagnose_design(simulations_df = simulations)
+#'
+#' }
+
+#' @importFrom stats setNames
+#' @importFrom utils head
+#' @export
+diagnose_design <-
```

Comment:
If we want this function to work with either lists of designs or simuation dfs, it should probably be generic instead. Split into diagnose_design.data.frame, which has all logic from the simulation forward, and diagnose_design.default, which creates simulations from designs/lists of designs, then passes that to diagnose_design.data.frame.